lib/async_hooks.js

```diff
@@ -98,24 +103,19 @@ class AsyncHook {
     if (hooks_array.includes(this))
       return;
 
-    if (!setupHooksCalled) {
-      setupHooksCalled = true;
-      // Setup the callbacks that node::AsyncWrap will call when there are
-      // hooks to process. They use the same functions as the JS embedder API.
-      async_wrap.setupHooks({ init,
-                              before: emitBeforeN,
-                              after: emitAfterN,
-                              destroy: emitDestroyN });
-    }
-
     // createHook() has already enforced that the callbacks are all functions,
     // so here simply increment the count of whether each callbacks exists or
     // not.
     hook_fields[kInit] += +!!this[init_symbol];
     hook_fields[kBefore] += +!!this[before_symbol];
     hook_fields[kAfter] += +!!this[after_symbol];
     hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hook_fields[kTotals]++;
```

Comment:
> I doubt it's a common case. Honestly I'm looking for what is the most expected behavior, but since creating a new hook w/o any callbacks isn't an expected behavior not sure what to say. Which behavior would you expect?

I will expect `async_hooks` to have the smallest performance impact possible. This includes not enabling promise hooks when there are no hooks.

> Personally I like this solution more because the logic is much simpler.

Yeah, that is the one reason why I don't like the additional checks. But I think especially in `async_hooks` performance is preferred over simplicity. It is not that difficult to check:

```js
hook_fields[kTotal] += +(
  !!this[init_symbol] || !!this[before_symbol] ||
  !!this[after_symbol] || !!this[destroy_symbol]
)
```